File: merge_datasets.js
// ============================================================
// Script merge all CSV files into one dataset
// Handle quoted CSV properly
// ============================================================
const fs = require('fs');
const path = require('path');
const readline = require('readline');

const baseDir = 'c:/CollectionData/Dataset';
const outputDir = 'c:/CollectionData';

// Source files
const files = [
  { file: 'Dataset DATN - Dataset - Concung.csv', source: 'concung' },
  { file: 'Dataset DATN - Dataset - Emart.csv', source: 'emart' },
  { file: 'Dataset DATN - Dataset - Guardian.csv', source: 'guardian' },
  { file: 'Dataset DATN - Dataset - NhaSachPhuongNam.csv', source: 'nhasachphuongnam' },
  { file: 'Dataset DATN - Dataset - SuperSports.csv', source: 'supersports' },
  { file: 'Dataset DATN - Dataset - Tiki.csv', source: 'tiki' },
  { file: 'Ecommerce Dataset - Dataset - Canifa.csv', source: 'canifa' },
  { file: 'Ecommerce Dataset - Dataset - Hòa Phát.csv', source: 'hoaphat' },
  { file: 'Ecommerce Dataset - Dataset - Nội Thất.csv', source: 'noithat' },
  { file: 'Ecommerce Dataset - Dataset - Nhà Xinh.csv', source: 'nhaxinh' },
  { file: 'Ecommerce Dataset - Dataset - TGDD.csv', source: 'tgdd' },
  { file: 'Ecommerce Dataset - Dataset - Yody.csv', source: 'yody' },
];

const headerCount = 23;
const headers = 'product_id,product_name,description,category_name,category_slug,category_url,category_image_url,brand,price,original_price,currency,stock,rating,reviews_count,thumbnail_url,image_urls,tags,color,size,material,product_url,source,searchable_text';

// Parse CSV line properly (handle quoted commas)
function parseCSVLine(line) {
  const result = [];
  let current = '';
  let inQuotes = false;
  
  for (let i = 0; i < line.length; i++) {
    const char = line[i];
    
    if (char === '"') {
      if (inQuotes && line[i + 1] === '"') {
        current += '"';
        i++;
      } else {
        inQuotes = !inQuotes;
      }
    } else if (char === ',' && !inQuotes) {
      result.push(current);
      current = '';
    } else {
      current += char;
    }
  }
  result.push(current);
  
  return result;
}

// Escape CSV value
function escapeCSV(value) {
  if (value == null || value === undefined) return '';
  const str = String(value);
  if (str.includes(',') || str.includes('"') || str.includes('\n') || str.includes('\r')) {
    return '"' + str.replace(/"/g, '""') + '"';
  }
  return str;
}

function processFile(filePath, source, startId) {
  return new Promise((resolve, reject) => {
    const rows = [];
    let globalProductId = startId;
    
    const rl = readline.createInterface({
      input: fs.createReadStream(filePath, { encoding: 'utf8' }),
      crlfDelay: Infinity
    });
    
    let isFirstLine = true;
    
    rl.on('line', (line) => {
      if (isFirstLine) {
        isFirstLine = false;
        return;
      }
      
      if (line.trim()) {
        const values = parseCSVLine(line);
        
        // Pad with empty values if needed
        while (values.length < headerCount) {
          values.push('');
        }
        
        // Update product_id (column 0)
        values[0] = globalProductId++;
        
        // Update source (column 21)
        if (values.length > 21) {
          values[21] = source;
        }
        
        // Escape all values
        const escapedValues = values.map((v, i) => {
          if (i === 0) return v; // product_id no escape
          return escapeCSV(v);
        });
        
        rows.push(escapedValues.join(','));
      }
    });
    
    rl.on('close', () => resolve({ rows, newStartId: globalProductId }));
    rl.on('error', reject);
  });
}

async function main() {
  console.log('=== MERGING DATASETS ===\n');
  
  let globalProductId = 1;
  const allRows = [headers];
  
  for (const { file, source } of files) {
    const filePath = path.join(baseDir, file);
    console.log(`Processing: ${file}`);
    
    try {
      const result = await processFile(filePath, source, globalProductId);
      allRows.push(...result.rows);
      
      console.log(`  - Processed ${result.rows.length} rows`);
      globalProductId = result.newStartId;
    } catch (e) {
      console.log(`  - Error: ${e.message}`);
    }
  }
  
  // Write output with BOM for Excel UTF-8 compatibility
  const outputPath = path.join(outputDir, 'Merged_Ecommerce_Dataset.csv');
  const outputContent = allRows.join('\r\n');
  fs.writeFileSync(outputPath, '\ufeff' + outputContent, 'utf8');
  
  console.log('\n=== SUMMARY ===');
  console.log(`Total files merged: ${files.length}`);
  console.log(`Total products: ${allRows.length - 1}`);
  console.log(`Output file: ${outputPath}`);
  console.log(`File size: ${(fs.statSync(outputPath).size / 1024 / 1024).toFixed(2)} MB`);
}

main().catch(console.error);

File: merge_datasets.py
# Script merge all CSV files - Using Pandas for proper CSV handling
import pandas as pd
import os
import warnings
warnings.filterwarnings('ignore')

base_dir = 'c:/CollectionData/Dataset'
output_dir = 'c:/CollectionData'

files = [
    ('Dataset DATN - Dataset - Concung.csv', 'concung'),
    ('Dataset DATN - Dataset - Emart.csv', 'emart'),
    ('Dataset DATN - Dataset - Guardian.csv', 'guardian'),
    ('Dataset DATN - Dataset - NhaSachPhuongNam.csv', 'nhasachphuongnam'),
    ('Dataset DATN - Dataset - SuperSports.csv', 'supersports'),
    ('Dataset DATN - Dataset - Tiki.csv', 'tiki'),
    ('Ecommerce Dataset - Dataset - Canifa.csv', 'canifa'),
    ('Ecommerce Dataset - Dataset - Hòa Phát.csv', 'hoaphat'),
    ('Ecommerce Dataset - Dataset - Nội Thất.csv', 'noithat'),
    ('Ecommerce Dataset - Dataset - Nhà Xinh.csv', 'nhaxinh'),
    ('Ecommerce Dataset - Dataset - TGDD.csv', 'tgdd'),
    ('Ecommerce Dataset - Dataset - Yody.csv', 'yody'),
]

# Define expected columns
columns = [
    'product_id', 'product_name', 'description', 'category_name', 'category_slug',
    'category_url', 'category_image_url', 'brand', 'price', 'original_price',
    'currency', 'stock', 'rating', 'reviews_count', 'thumbnail_url', 'image_urls',
    'tags', 'color', 'size', 'material', 'product_url', 'source', 'searchable_text'
]

print('=== MERGING DATASETS (Pandas) ===\n')

all_dfs = []

for file, source in files:
    file_path = os.path.join(base_dir, file)
    print(f'Processing: {file}')
    
    try:
        # Try different encodings and parsing options
        for encoding in ['utf-8', 'utf-8-sig', 'latin-1', 'cp1252']:
            try:
                # Use python engine for more flexible parsing
                df = pd.read_csv(
                    file_path,
                    encoding=encoding,
                    engine='python',
                    on_bad_lines='skip',
                    quotechar='"',
                    escapechar='\\'
                )
                break
            except:
                continue
        
        # Check if we got columns
        if df.shape[1] < 5:
            print(f'  ⚠️ Skipped - not enough columns ({df.shape[1]})')
            continue
        
        # Rename columns if needed
        if df.shape[1] >= 23:
            df = df.iloc[:, :23]
        
        # Ensure we have 23 columns
        while df.shape[1] < 23:
            df[f'col_{df.shape[1]}'] = ''
        
        df = df.iloc[:, :23]
        df.columns = columns
        
        # Update source
        df['source'] = source
        
        # Add to list
        all_dfs.append(df)
        print(f'  ✅ Processed {len(df)} rows')
        
    except Exception as e:
        print(f'  ❌ Error: {str(e)[:50]}')

# Concatenate all dataframes
if all_dfs:
    merged_df = pd.concat(all_dfs, ignore_index=True)
    
    # Reset product_id
    merged_df['product_id'] = range(1, len(merged_df) + 1)
    
    # Save
    output_path = os.path.join(output_dir, 'Merged_Ecommerce_Dataset.csv')
    merged_df.to_csv(output_path, index=False, encoding='utf-8-sig')
    
    print('\n=== SUMMARY ===')
    print(f'Total files merged: {len(all_dfs)}')
    print(f'Total products: {len(merged_df):,}')
    print(f'Output file: {output_path}')
    print(f'File size: {os.path.getsize(output_path) / 1024 / 1024:.2f} MB')
else:
    print('❌ No data to merge!')
